report_dynamic/models/report_dynamic.py

```diff
@@ -0,0 +1,242 @@
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class ReportDynamic(models.Model):
+    _name = "report.dynamic"
+    _description = "Dynamically create reports"
+
+    name = fields.Char(required=True)
+    model_id = fields.Many2one("ir.model")
+    # Inform the user about configured model_id
+    # in template
+    model_name = fields.Char(related="model_id.name")
+    model_model = fields.Char(related="model_id.model", string="Tech name of model_id")
+    res_id = fields.Integer()
+    resource_ref = fields.Reference(
+        string="Target record",
+        selection="_selection_target_model",
+        compute="_compute_resource_ref",
+        inverse="_inverse_resource_ref",
+    )
+    wrapper_report_id = fields.Many2one("ir.ui.view", domain="[('type', '=', 'qweb')]")
+    template_id = fields.Many2one(
+        "report.dynamic",
+        domain="[('is_template', '=', True)]",
+        copy=False,
+        default=lambda self: self.env.company.external_report_layout_id,
+    )
+    report_ids = fields.One2many(
+        "report.dynamic",
+        "template_id",
+        domain="[('is_template', '=', False)]",
+        copy=False,
+    )
+    documentation = fields.Text(default="Documentation placeholder", readonly=True)
+    condition_domain_global = fields.Char(
+        string="Global domain condition", default="[]"
+    )
+    active = fields.Boolean(default=True)
+    is_template = fields.Boolean(default=False)
+    lock_date = fields.Date()
+    field_ids = fields.Many2many(
+        "ir.model.fields", "contextual_field_rel", "contextual_id", "field_id", "Fields"
+    )
+    window_action_exists = fields.Boolean(compute="_compute_window_action_exists")
+
+    @api.model
+    def _selection_target_model(self):
+        models = self.env["ir.model"].search([])
+        return [(model.model, model.name) for model in models]
```

Comment:
check here if there's an efficient way to present only those models that at least one active record exists in the db.

Reply:
I think just simply get the distinct model ids or names from the report dynamic table first, and then use them as filter criteria in this search domain.

Reply:
Of course you could also refactor this selection field to a many2one to ir.model which might make it a bit easier all in all but not sure if its worth it

Reply:
yes, that's what I am attempting to do, to get non transient models with count > 0. Will resume working on this soon, most probably.

Reply:
Yea in one step thats difficult, you could do that with sql but thats definitely not worth it:) if its two steps you can just unique them in python and the ones in report dynamic will be non transient anyway

Reply:
Probably replacing the reference field with a many2one to ir.model, and another many2one to ir.model.model, will make this easier. Something to consider in the future.

Reply:
I would start by filtering out non real models. Then you can add a TODO to improve it.

Reply:
```sql
SELECT 
  nspname AS schemaname,relname,reltuples
FROM pg_class C
LEFT JOIN pg_namespace N ON (N.oid = C.relnamespace)
WHERE 
  nspname NOT IN ('pg_catalog', 'information_schema') AND
  relkind='r' 
ORDER BY reltuples DESC;
```

the above takes the amount of tuples from the statistics (created by autovacuum) - but I consider only showing models with records pretty much a non-feature. I'd want to be able to prepare a report even if I haven't created records already, and also installing some module using this to create a report would fail if the model doesn't have records